### PR TITLE
[corlib] Improve reflection interface definitions

### PR DESCRIPTION
**Pull request by nsivov**
It's important to have compatible interface definitions for COM interop to work correctly. Particular case this was tested with is WiX installer tool running with wine + wine-mono. It's using mscoree API to do the following:

- create domain -> load assembly and get _Assembly interface for it -> locate type as _Type -> locate method info as _MethodInfo -> call Invoke() on it.

Current _MethodInfo and _Type seem to be roughly arranged in alphabetical order.

### CODE REVIEW

File: mcs/class/corlib/ReferenceSources/Type.cs [hunk 1]
```diff
@@ -34,10 +34,40 @@
 namespace System
 {
 	[Serializable]
+	[ComVisible (true)]
+	[ComDefaultInterfaceAttribute (typeof (_Type))]
```
Comment (rjvdboon) [inline mcs/class/corlib/ReferenceSources/Type.cs]: Should this line be guarded by `#if !MOBILE`?

File: mcs/class/corlib/System.Reflection/MethodInfo.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+using System.Runtime.InteropServices;
+
+namespace System.Reflection {
+
+	[ComVisible(true)]
+	[ComDefaultInterface(typeof(_MethodInfo))]
```
Comment (rjvdboon) [inline mcs/class/corlib/System.Reflection/MethodInfo.cs]: Should this line be guarded by `#if !MOBILE`?

File: mcs/class/corlib/ReferenceSources/Type.cs (hunk 1 above)
Comment (nsivov) [reply mcs/class/corlib/ReferenceSources/Type.cs]: It's not guarded in case of Assembly (_Assembly), so I'm not sure it matters.

**Comment by nsivov** [comment]
I'd appreciate any feedback on this one.

**Comment by akoeplinger** [comment]
It seems to cause a test issue in the interpreter configurations when FullAOT is used:

```
gshared.exe

Unhandled Exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ExecutionEngineException: Attempting to JIT compile method '(wrapper managed-to-native) System.Type object:GetType (object)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.

  at Tests+CGetType.gettype2[T] (T t) [0x00001] in <628322f0c5b0473196a132fcf2ed297d>:0 
  at Tests.test_0_constrained_gettype () [0x00051] in <628322f0c5b0473196a132fcf2ed297d>:0 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00083] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
  at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
  at TestDriver.RunTests (System.Type type, System.String[] args, TestDriverReporter reporter) [0x003b9] in <d02825add33c4ae4a1ba3e398253e8b8>:0 
  at TestDriver.RunTests (System.Type type, System.String[] args) [0x00001] in <d02825add33c4ae4a1ba3e398253e8b8>:0 
  at Tests.Main (System.String[] args) [0x00001] in <628322f0c5b0473196a132fcf2ed297d>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ExecutionEngineException: Attempting to JIT compile method '(wrapper managed-to-native) System.Type object:GetType (object)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.

  at Tests+CGetType.gettype2[T] (T t) [0x00001] in <628322f0c5b0473196a132fcf2ed297d>:0 
  at Tests.test_0_constrained_gettype () [0x00051] in <628322f0c5b0473196a132fcf2ed297d>:0 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00083] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
  at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <41702dbbcf5f4fec8fa23d4c11989bd7>:0 
  at TestDriver.RunTests (System.Type type, System.String[] args, TestDriverReporter reporter) [0x003b9] in <d02825add33c4ae4a1ba3e398253e8b8>:0 
  at TestDriver.RunTests (System.Type type, System.String[] args) [0x00001] in <d02825add33c4ae4a1ba3e398253e8b8>:0 
  at Tests.Main (System.String[] args) [0x00001] in <628322f0c5b0473196a132fcf2ed297d>:0 
Makefile:2916: recipe for target 'fullaotcheck' failed
make[1]: *** [fullaotcheck] Error 1
make[1]: Leaving directory '/mnt/jenkins/workspace/test-mono-pull-request-interpreter/mono/mini'
Makefile:2912: recipe for target 'fullaotmixedcheck' failed
make: *** [fullaotmixedcheck] Error 2
make: Leaving directory '/mnt/jenkins/workspace/test-mono-pull-request-interpreter/mono/mini'
/mnt/jenkins/workspace/test-mono-pull-request-interpreter/scripts/ci/babysitter: Test suite terminated with code 2, and suite cannot report test case data. Halting.

*** end(33): fullaotmixed-regression: [43mUnstable[0m
```

**Comment by nsivov** [comment]
Thank you, I'll take a look.

File: mcs/class/corlib/ReferenceSources/Type.cs (hunk 1 above)
Comment (marek-safar) [reply mcs/class/corlib/ReferenceSources/Type.cs]: It should be

File: mcs/class/corlib/System.Reflection/MethodInfo.cs (hunk 2 above)
Comment (marek-safar) [reply mcs/class/corlib/System.Reflection/MethodInfo.cs]: Yes

**Comment by nsivov** [comment]
I was able to reproduce this test failure locally back then. Will try to understand what's going on with AOT case.

**Comment by marek-safar** [comment]
@monojenkins commit apidiff

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by nsivov** [comment]
Failure with gshared.exe is caused by `partial class MethodInfo : _MethodInfo` change, removing inheritance from _MethodInfo makes it work again.

**Comment by nsivov** [comment]
@marek-safar pushed something that seems to fix it locally, let's see how it goes

**Comment by marek-safar** [comment]
@monojenkins commit csproj and apidiff

**Pull request merged**